stdlib/event/gui.lua

```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
 
     if event.element and event.element.valid then
         event.tick = event.tick or game.tick
-        for gui_element_pattern, handler in pairs(Event._registry[event.name]) do
+        for gui_element_pattern, handlers in pairs(Gui._registry[event.name]) do
             if event.element and event.element.valid then
                 local match_str = event.element.name:match(gui_element_pattern)
                 if match_str then
                     event.match = match_str
                     event.state = event.name == defines.events.on_gui_checked_state_changed and event.element.state or nil
                     event.text = event.name == defines.events.on_gui_text_changed and event.element.text or nil
-                    setmetatable(event, { __index = { _handler = handler } })
-                    local success, err = pcall(handler, event)
-                    if not success then
-                        game.print(err)
+                    local metahandler = {}
+                    setmetatable(event, {__index = metahandler})
+                    for _, handler in ipairs(handlers) do
+                        metahandler._handler = handler
+                        local success, err = pcall(handler, event)
+                        if not success then
+                            if game and game.print then
+                                game.print(err)
+                            elseif log then
+                                log(err)
+                            else -- jic...
+                                print(err)
+                            end
+                        end
                     end
                 end
             end
         end
     end
 end
 
---- Removes the handler with matching gui element pattern from the event.
--- @tparam defines.events event_id valid values are `defines.events.on_gui_*` from @{defines.events}
--- @tparam string gui_element_pattern the name or string regular expression for a handler to remove
+--- Registers a function for a given event and matching gui element pattern.
+-- @tparam defines.events event_ids valid values are `defines.events.on_gui_*` from @{defines.events},
+-- or a list of them.
+-- @tparam string gui_element_pattern the name or string regular expression to match the gui element
+-- @tparam function handler the function to call when the event is triggered
 -- @return (<span class="types">@{Gui}</span>)
-function Gui.remove(event_id, gui_element_pattern)
-    fail_if_missing(event_id, "missing event argument")
+function Gui.register(event_ids, gui_element_pattern, handler)
+    fail_if_missing(event_ids, "missing event name argument")
+    fail_if_missing(gui_element_pattern, "missing gui name or pattern argument")
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.register(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
 
     if type(gui_element_pattern) ~= "string" then
```

Comment:
This check should go above the recursive loop because,    fail first :)   however if the above fail_if_missing change is added this check can go the way of the Dodo bird

Reply:
ACK, also downstream.